linera-service/src/exporter/runloops/block_processor/mod.rs

```
@@ -0,0 +1,326 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{future::IntoFuture, time::Duration};
+
+use linera_storage::Storage;
+use tokio::{
+    sync::mpsc::{UnboundedReceiver, UnboundedSender},
+    time::{interval, MissedTickBehavior},
+};
+
+use crate::{
+    common::{BlockId, ExporterError},
+    runloops::block_processor::walker::Walker,
+    storage::BlockProcessorStorage,
+};
+
+mod walker;
+
+pub(super) struct BlockProcessor<T>
+where
+    T: Storage + Clone + Send + Sync + 'static,
+{
+    storage: BlockProcessorStorage<T>,
+    queue_rear: UnboundedSender<BlockId>,
+    queue_front: UnboundedReceiver<BlockId>,
+}
+
+impl<T> BlockProcessor<T>
+where
+    T: Storage + Clone + Send + Sync + 'static,
+{
+    pub(super) fn new(
+        storage: BlockProcessorStorage<T>,
+        queue_rear: UnboundedSender<BlockId>,
+        queue_front: UnboundedReceiver<BlockId>,
+    ) -> Self {
+        Self {
+            storage,
+            queue_rear,
+            queue_front,
+        }
+    }
+
+    pub(super) async fn run_with_shutdown<F>(
+        &mut self,
+        shutdown_signal: F,
+        persistence_period: u32,
+    ) -> Result<(), ExporterError>
+    where
+        F: IntoFuture<Output = ()>,
+    {
+        let shutdown_signal_future = shutdown_signal.into_future();
+        let mut pinned_shutdown_signal = Box::pin(shutdown_signal_future);
+
+        let mut interval = interval(Duration::from_millis(persistence_period.into()));
+        interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+
+        loop {
+            tokio::select! {
+
+                biased;
+
+                _ = &mut pinned_shutdown_signal => break,
+
+                _ = interval.tick() => self.storage.save().await?,
+
+                Some(next_block_notification) = self.queue_front.recv() => {
+                    let walker = Walker::new(&mut self.storage);
+                    if let Err(ExporterError::ViewError(_)) = walker.walk(next_block_notification).await {
```

Are the other error variants logged anywhere? It seems like we would silently drop the result and continue "business as usual" but that will most likely mean some chains will be blocked forever. If we fail "export block 10" we will also fail to index subsequent blocks. 

Log other variants as warning - include the block's ID (hash) and the error.

---

I have gone through all the errors, they are either safely skippable or unreachable

---

Log other variants as warning - include the block's ID (hash) and the error.